libbpf-tools/memleak.c

```diff
@@ -0,0 +1,972 @@
+		}
+
+		// when the stack_id exists in the allocs array,
+		//   increment size with alloc_info.size
+		bool stack_exists = false;
+
+		for (size_t i = 0; !stack_exists && i < nr_allocs; ++i) {
+			struct allocation *alloc = &allocs[i];
+
+			if (alloc->stack_id == alloc_info.stack_id) {
+				alloc->size += alloc_info.size;
+				alloc->count++;
+
+				stack_exists = true;
+				break;
+			}
+		}
+
+		if (stack_exists)
+			continue;
+
+		// when the stack_id does not exist in the allocs array,
+		//   create a new entry in the array
+		const struct allocation alloc = {
+			.stack_id = alloc_info.stack_id,
+			.size = alloc_info.size,
+			.count = 1
+		};
+
+		memcpy(&allocs[nr_allocs], &alloc, sizeof(alloc));
+		nr_allocs++;
+	}
+
+	// sort the allocs array in descending order
+	qsort(allocs, nr_allocs, sizeof(allocs[0]), alloc_size_compare);
+
+	// get min of allocs we stored vs the top N requested stacks
+	nr_allocs = nr_allocs < env.top_stacks ? nr_allocs : env.top_stacks;
+
+	printf("[%d:%d:%d] Top %zu stacks with outstanding allocations:\n",
+			tm->tm_hour, tm->tm_min, tm->tm_sec, nr_allocs);
+
+	print_stack_frames(allocs, nr_allocs, stack_traces_fd);
+
+	return 0;
+}
+
+int print_outstanding_combined_allocs(int combined_allocs_fd, int stack_traces_fd)
+{
+	time_t t = time(NULL);
+	struct tm *tm = localtime(&t);
+
+	size_t nr_allocs = 0;
+
+	// for each stack_id "curr_key" and union combined_alloc_info "alloc"
+	// in bpf_map "combined_allocs"
+	for (uint64_t prev_key = 0, curr_key = 0;; prev_key = curr_key) {
+		union combined_alloc_info combined_alloc_info;
+		memset(&combined_alloc_info, 0, sizeof(combined_alloc_info));
+
+		if (bpf_map_get_next_key(combined_allocs_fd, &prev_key, &curr_key)) {
+			if (errno == ENOENT) {
+				break; // no more keys, done
+			}
+
+			perror("map get next key error");
+
+			return -errno;
+		}
+
+		if (bpf_map_lookup_elem(combined_allocs_fd, &curr_key, &combined_alloc_info)) {
+			if (errno == ENOENT)
+				continue;
+
+			perror("map lookup error");
+
+			return -errno;
+		}
+
+		const struct allocation alloc = {
+			.stack_id = curr_key,
+			.size = combined_alloc_info.total_size,
+			.count = combined_alloc_info.number_of_allocs
+		};
+
+		memcpy(&allocs[nr_allocs], &alloc, sizeof(alloc));
+		nr_allocs++;
+	}
+
+	qsort(allocs, nr_allocs, sizeof(allocs[0]), alloc_size_compare);
+
+	// get min of allocs we stored vs the top N requested stacks
+	nr_allocs = nr_allocs < env.top_stacks ? nr_allocs : env.top_stacks;
+
+	printf("[%d:%d:%d] Top %zu stacks with outstanding allocations:\n",
+			tm->tm_hour, tm->tm_min, tm->tm_sec, nr_allocs);
+
+	print_stack_frames(allocs, nr_allocs, stack_traces_fd);
+
+	return 0;
+}
+
+bool has_kernel_node_tracepoints()
+{
+	return tracepoint_exists("kmem", "kmalloc_node") &&
+		tracepoint_exists("kmem", "kmem_cache_alloc_node");
+}
+
+int disable_kernel_node_tracepoints(struct memleak_bpf *skel)
+{
+	if (bpf_program__set_autoload(skel->progs.tracepoint__kmalloc_node, false) ||
```

Comment:
set_autoload can't fail if called before skeleton load, so unnecessary error checking here and in other places